src/main/java/dp/api/dataset/DatasetAPIClient.java

```diff
@@ -0,0 +1,378 @@
+package dp.api.dataset;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import dp.api.dataset.exception.BadRequestException;
+import dp.api.dataset.exception.DatasetAPIException;
+import dp.api.dataset.exception.DatasetNotFoundException;
+import dp.api.dataset.exception.InstanceNotFoundException;
+import dp.api.dataset.exception.UnexpectedResponseException;
+import dp.api.dataset.model.Dataset;
+import dp.api.dataset.model.DatasetResponse;
+import dp.api.dataset.model.DatasetVersion;
+import dp.api.dataset.model.Instance;
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpStatus;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpEntityEnclosingRequestBase;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.client.methods.HttpPut;
+import org.apache.http.client.methods.HttpRequestBase;
+import org.apache.http.client.utils.URIBuilder;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.Args;
+import org.apache.http.util.EntityUtils;
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URI;
+import java.net.URISyntaxException;
+
+/**
+ * HTTP client for the dataset API.
+ */
+public class DatasetAPIClient implements DatasetClient {
```

Comment:
Can this class extend closable as well? as CloseableHTTPClient is a internal resource.

Reply:
done